procrastinate/psycopg2_connector.py

```diff
@@ -124,26 +155,42 @@ def _wrap_json(self, arguments: Dict[str, Any]):
 
     @contextlib.contextmanager
     def _connection(self) -> psycopg2.extensions.connection:
+        # in case of an admin shutdown (Postgres error code 57P01) we do not
+        # rollback the connection or put the connection back to the pool as
+        # this will cause a psycopg2.InterfaceError exception
         try:
-            with self._pool.getconn() as connection:
+            connection = self._pool.getconn()
+            try:
                 yield connection
-        finally:
+            except psycopg2.errors.AdminShutdown:
+                raise
```

Comment:
So IIUC, the whole thing is that in case of AdminShutdown, we must avoid rolling back, right ?

Reply:
Yep, see the command above. And we also avoid putting the connection back to the pool.

Reply:
Ah but you don't do this for `except Exception` too, is it voluntary ?

How do you tell psycopg2 to stop keeping track of those connections ?

Reply:
> Ah but you don't do this for `except Exception` too, is it voluntary ?

I now call `putconn` after rollback.